src/js/actions/documents.js

```diff
@@ -898,18 +883,54 @@ define(function (require, exports) {
         }
 
         var newVisibility = !document.smartGuidesVisible,
-            dispatchPromise = this.dispatchAsync(events.document.GUIDES_VISIBILITY_CHANGED,
-                { documentID: document.id, smartGuidesVisible: newVisibility });
-
-        var playObject = documentLib.setSmartGuidesVisibility(newVisibility),
+            playObject = documentLib.setSmartGuidesVisibility(newVisibility),
             playPromise = descriptor.playObject(playObject);
 
-        return Promise.join(dispatchPromise, playPromise);
+        return playPromise
+            .bind(this)
+            .then(function () {
+                return _dispatchGuidesVisibilityEvent.call(this, document);
+            });
     };
     toggleSmartGuidesVisibility.action = {
         reads: [locks.JS_DOC, locks.JS_APP],
         writes: [locks.JS_DOC, locks.PS_DOC]
     };
+    
+    /**
+     * A helper function to get the visibility state of guides and smart guides from PS and dispatch
+     * GUIDES_VISIBILITY_CHANGED event to update the related menu items.
+     * 
+     * @param  {Document} document
+     * @return {Promise}
+     */
+    var _dispatchGuidesVisibilityEvent = function (document) {
+        var guidesPromise = descriptor.playObject(documentLib.getGuidesVisibility()),
```

Comment:
It would be better if you used batchPlayObjects here so we don't make separate calls into PS.